Add option to hide help based on long/short

stevepentland opened the pull request:
Relates to #1064 

I've added the enum variants and related setting methods to provide the functionality. However, the usage of [Parser::use_long_help()](https://github.com/kbknapp/clap-rs/blob/master/src/app/parser.rs#L1501) is causing issues. 

It seems that the linked method will not allow `use_long` to be `true` unless there is actually a long help value defined for the `Arg`. However, I noticed in [help.rs](https://github.com/kbknapp/clap-rs/blob/master/src/app/help.rs#L445) that we would choose to use the desired long or short help based on what was specified during the run and the extra logic does not seem like it is really needed. However when I bypassed the code in `parser.rs`, there were some failing tests.

@kbknapp do you have any suggestions?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kbknapp/clap-rs/1235)
<!-- Reviewable:end -->

kbknapp commented:
Ah ok I see now. I'm investigating a little more on my end to see what the issue may be, I'll let you know what I find.

kbknapp commented:
Ok, so I see what's going on now (at least the `Parser::use_long_help`). It was meant to be able to say, "If no args defined a `long_help` just go ahead and use the old short help style." So people's help messages didn't suddenly change.

i.e.

Old style when using `--help` and no `long_help` has been defined

```
test 1.3
Kevin K.
tests stuff

USAGE:
    test [OPTIONS] [arg1]

OPTIONS:
    -f, --flag            some flag
    -h, --help            Prints help information
        --option <opt>    some option
    -V, --version         Prints version information

ARGS:
    <arg1>    some pos arg
```

Without `Parser::use_long_help`, even without defining any `long_help`, when the user does `--help` the message changes to this:

```
test 1.3
Kevin K.
tests stuff

USAGE:
    test [OPTIONS] [arg1]

OPTIONS:
    -f, --flag            
            some flag

    -h, --help            
            Prints help information

        --option <opt>    
            some option

    -V, --version         
            Prints version information


ARGS:
    <arg1>    
            some pos arg
```

(note the help text being printed on the line below).

I'm personally fine with this change and dropping `Parser::use_long_help` entirely and if the user uses `--help` they see the long message, otherwise they see the short message with `-h`. I've never considered the *format* of the output of the help messages to part of the public API, only the *correctness*. It's had small formatting changes and such throughout v1 and v2.

However, it is a somewhat in your face stylistic change and would like to hear a few other opinions before making that particular change.

--- 

Back to the issue at hand, a simple way to keep the old functionality *and* fix this issue is to not only check if `long_help.is_some()` on all args, but also `.is_set(ArgSettings::HiddenLongHelp)`.

stevepentland commented:
> Back to the issue at hand, a simple way to keep the old functionality and fix this issue is to not only check if `long_help.is_some()` on all args, but also `.is_set(ArgSettings::HiddenLongHelp)`.

It does work, but it also brings up another potential issue: While there is no change when `hidden_long_help` has not been set, once it has been, the output suddenly changes to the style you outlined in your comment. This may catch some people off guard if they're adding `hidden_long_help` and have no previously defined `long_help` values as now there are 2 different styles for `--help`.

kbknapp commented:
So long as it's documented I'm fine with it. I think the case to hide something from the long help only will be extremely rare, and already somewhat implies two different help messages (even if *style* isn't implied).

kbknapp commented:
@stevepentland let me know once you're good with it and I'll do the final review

stevepentland commented:
@kbknapp will do, should be soon. Just gotten side tracked

stevepentland commented:
@kbknapp I've gotten them both working, updated the check, and added docs stating that using these options will change how `--help` displays. I also added tests.

kbknapp commented:
Looks good, thanks! :+1:
